### PR TITLE
Compatibility fixes/improvements for JSON/RPC filter polling

### LINKED ISSUE #640: eth_getFilterChanges returns a string containing serialized JSON array (not the JSON array)
### Description

If you install a block filter, using `eth_newBlockFilter`, or a log filter, using `eth_newFilter`, then call `eth_getFilterChanges` you get a result such as:

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "result": "[\"0x3b6c51a0447f94cbcafa3cf8ae05721b972f18d6aafa2be69da31712f38699ef\",\"0x941de60877552e0559049d7e60ce53cbdffdf098ad039d786fa76aedbb63df26\",\"0x1cfcad6fa8a4d33d586817db2aac7ad7de8e4b825ba44a166f80f6d94d665661\",\"0xde6ccf731581c9b34901aea3487ec0b692bb17d3f2a3eebe99d824fb60c921bf\",\"0x73dcb8107af6d7bc882e528549f32c7f12f01fc0b8db116b92312fba75d72b65\",\"0x0d56fb2ba64770007132b1f56c662c547666735b6bebb2a9f8e521bfdd8d8e9d\"]"
}
```

You can see 

### Your environment

- Edge version: v0.4.1
- Dockerized linux (Ubuntu 18.04 base).

#### For blocks

Request 1: `eth_newBlockFilter`

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "eth_newBlockFilter"
}
```

Response 1: `eth_newBlockFilter`

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "result": "2ee73d8b-5d37-46e1-b743-511b0e54a55d"
}
```

Request 2: `eth_getFilterChanges`

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "eth_getFilterChanges",
    "params": ["2ee73d8b-5d37-46e1-b743-511b0e54a55d"]
}
```

Response 2: `eth_getFilterChanges`

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "result": "[\"0x2e2fef24eea91c245e32cd2793b668a2b18b6202914e3e62e029709efca5c48c\"]"
}
```

#### For Logs

Request 1: `eth_newFilter`

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "eth_newFilter",
    "params": [
        {
            "fromBlock": "0",
            "topics": ["0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"]
        }
    ]
}
```

Response 1:

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "result": "c2c92aa7-75ac-4d98-b392-00ccd10a4e73"
}
```

Request 2: `eth_getFilterLogs` - note this is **GOOD**

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "eth_getFilterLogs",
    "params": ["c2c92aa7-75ac-4d98-b392-00ccd10a4e73"]
}
```

Response 2: - note this is **GOOD**

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "result": [
        {
            "address": "0x432346fc656AbA8b1a4111374A9bDf56Df05d8d9",
            "topics": [
                "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
                "0x0000000000000000000000000000000000000000000000000000000000000000",
                "0x0000000000000000000000000151d512928dd4e1a0a0c9f3c125bf0380e6d587"
            ],
            "data": "0x0000000000000000000000000000000000000000033b2e3c9fd0803ce8000000",
            "blockNumber": "0x942",
            "transactionHash": "0xca1b4ff89454de929dfb1b12e264bfdedc9d09ef159b66b0699c4fe12952b945",
            "transactionIndex": "0x0",
            "blockHash": "0x41bf24b6a1d6f4af222558b57bf9b14759507dd75539e0bea8e6177242c955d2",
            "logIndex": "0x2",
            "removed": false
        }
    ]
}
```

Request 3: `eth_getFilterChanges`

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "method": "eth_getFilterChanges",
    "params": ["07f14e93-7a10-446d-8c12-e08e482abc05"]
}
```

Response 3: - see here it is a string (not an array)

```json
{
    "jsonrpc": "2.0",
    "id": "1",
    "result": "[{\"address\":\"0x432346fc656AbA8b1a4111374A9bDf56Df05d8d9\",\"topics\":[\"0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef\",\"0x0000000000000000000000000000000000000000000000000000000000000000\",\"0x0000000000000000000000000151d512928dd4e1a0a0c9f3c125bf0380e6d587\"],\"data\":\"0x0000000000000000000000000000000000000000000000000000000000003039\",\"blockNumber\":\"0xa17\",\"transactionHash\":\"0x0be6eb34b8c03342dd8134fd22cbeaa2a0b27bbd3d8da4f6fcfe64cf624c7a29\",\"transactionIndex\":\"0x0\",\"blockHash\":\"0xe697e957bba8a04e79616dc266c683fbbed020caa501ac4f07e8ccae9b6a32b0\",\"logIndex\":\"0x0\",\"removed\":false}]"
}
```

### Expected behaviour

The result should not be an array, not a string containing a serialized JSON array.

### Actual behaviour

It's an array directly in the JSON/RPC `result`.

### Logs

Easy to recreate.

Note: The unit test does not catch this, as the Go JSON/RPC wrapper used in the test masks this issue, by unmarshalling the array:
https://github.com/umbracle/ethgo/blob/8326bf607acf2ae4893128cacf205c67261cc411/jsonrpc/eth.go#L74-L86

### Proposed solution

The problem appears to be specific to this area of code, because it's trying to handle either type of filter there could be two types of result.

https://github.com/0xPolygon/polygon-edge/blob/d34f55789b7a89ee4f594b6a0793d3fa37f92f6a/jsonrpc/filter_manager.go#L481-L503

~So I propose this code returns a `json.RawMessage` rather than `string`.~
> See discussion in #641 - this exposed a follow-on issue in the code building a JSON string array for empty arrays, here:
https://github.com/0xPolygon/polygon-edge/blob/d34f55789b7a89ee4f594b6a0793d3fa37f92f6a/jsonrpc/filter_manager.go#L131
> It seems there's code up-stream that handles `interface{}` - so there's no need for this extra serialization step - so updating the PR for that.

I will work to provide that proposal in code via a PR, as a follow-up to this issue.

### PR DESCRIPTION
**Pull request by peterbroadhurst**
# Description

A set of changes that improve consistency of the Edge JSON/RPC surface area for block/log filters, with other Ethereum JSON/RPC client implementations. These proposed changes vary from clear fixes, to other items that are more subjective. So I'm leaving this in draft while I work through a set of tests with event streaming framework ethconnect - which is part of the Hyperledger FireFly project.

- Fix for #640 
- Consistent `filter not found` error string
- Allowing filters to exist for more than one minute, when being actively used

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

If any users were depending on the invalid behavior of the JSON/RPC `eth_getFilterChanges` returning a string `result` (rather than an array).

~This needs to stay in draft until a discussion happens on how to handle https://github.com/umbracle/ethgo, which has code that depends on the incorrect behavior. Resulting in this test failure:~

_Now resolved via https://github.com/umbracle/ethgo/pull/209_

```
--- FAIL: TestNewFilter_Logs (5.17s)
    logs_test.go:84: 
                Error Trace:    /Users/pbroadhurst/dev/polygon/polygon-edge/e2e/logs_test.go:84
                Error:          Received unexpected error:
                                json: cannot unmarshal array into Go value of type string
                Test:           TestNewFilter_Logs
```

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

# Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

**Comment by peterbroadhurst** [comment]
I've found a follow-on issue, when a block filter returns an empty result set.
In this case we get `[""]` (rather than `[]`.
Will extend the PR to cover this case

File: jsonrpc/filter_manager.go [hunk 1]
```diff
@@ -17,7 +16,7 @@ import (
 )
 
 var (
-	ErrFilterDoesNotExists              = errors.New("filter does not exists")
```
Comment (peterbroadhurst) [inline jsonrpc/filter_manager.go:20]: I noticed this typo, and also have proposed returning a string containing the words `filter not found`.

This is a condition that clients need to be able to distinguish from other errors, and the Ethereum JSON/RPC does not have a formal standard for this. However, other Ethereum clients by convention contain this string. So to reduce complexities for developers integrating with multiple clients, I suggest following that convention.

Comment (0xAleksaOpacic) [reply jsonrpc/filter_manager.go:20]: Agree thank you

**Comment by zivkovicmilos** [comment]
Hey @peterbroadhurst,

Thank you for opening up a PR.

~Can you please explain what kind of functionality is expected from the filter response, that Edge is not providing?~
I just saw the opened issue where you detailed the behavior in Edge.

Also, I'd like to know what kind of problem exists with the `ethgo` package regarding this functionality. 
We are moving towards dropping this package completely from Edge in the future as it's been more of a burden to maintain than our actual JSON-RPC interface.

If you can describe the problem here with `ethgo`, I'd be able to get a PR out and a fix merged very quickly as I'm one of the collaborators on that repo.

**Comment by peterbroadhurst** [comment]
Thanks you @zivkovicmilos - I'm sorry for the slow reply here.

The problem in the `ethgo` package is here: https://github.com/umbracle/ethgo/blob/8326bf607acf2ae4893128cacf205c67261cc411/jsonrpc/eth.go#L74-L86

This is actually working around the behavior currently on the JSON/RPC interface for `eth_getFilterChanges`, by unmarhsaling the string response into `[]*ethgo.Log` within the client code.

So when the fix in this PR changes the JSON/RPC response to an array (rather than a string), the error occurs.

So the PR needed in `ethgo` is just to pass the `[]*ethgo.Log` interface directly to `e.c.Call()` (rather than the `string`).

I had mean to submit this as a PR to `ethgo`, and got distracted by other priorities in the past couple of days. Very sorry about that 🙇 . I will do that now.

**Comment by peterbroadhurst** [comment]
I've raised https://github.com/umbracle/ethgo/pull/209 - I'm going to check in a change while this PR is in draft, that updates the go.mod to redirect to that branch so I can check it allows the PR tests to pass.

**Comment by peterbroadhurst** [comment]
Sadly my attempt to test with my fork of `ethgo` failed, as the test run against the vendor directory (and `go mod vendor` didn't respect the `replace` directive it seems). So it might be necessary to work through https://github.com/umbracle/ethgo/pull/209 before I can re-test here.

**Comment by peterbroadhurst** [comment]
Ok - with https://github.com/umbracle/ethgo/pull/209 merged, I've updated the PR to pull in that commit of `ethgo`.

**Comment by peterbroadhurst** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by zivkovicmilos** [comment]
Hey @peterbroadhurst,

Can you please pull in the latest `develop` branch into this PR so we can go ahead and merge it? 🙏 

We've made some minor fixes in the meantime that you'll need to resolve.

File: jsonrpc/filter_manager.go [hunk 2]
```diff
@@ -17,7 +16,7 @@ import (
...
+	ErrFilterDoesNotExists              = errors.New("filter not found")
```
Comment (Kourin1996) [inline jsonrpc/filter_manager.go]: How about renaming variable name to `ErrFilterNotFound` as well?

Comment (peterbroadhurst) [reply jsonrpc/filter_manager.go]: Resolved along with the merge 👍

**Comment by peterbroadhurst** [comment]
Thanks @zivkovicmilos - I've completed the merge, and added the change suggested by @Kourin1996

**Review by zivkovicmilos — APPROVED** [review]
Looks good 💯 

Thank you for the contribution 🙏

**Review by Kourin1996 — APPROVED** [review]
Looks good to me. Thank you for the PR!

**Pull request merged**